docs/proposals/30-apiserver-proxy.md

```diff
@@ -0,0 +1,268 @@
+  name: api-gateway
+  namespace: istio-system
+spec:
+  selector:
+    istio: ingressgateway
+  servers:
+    ...
+---
+apiVersion: networking.istio.io/v1alpha3
+kind: VirtualService
+metadata:
+  name: api-routes
+  namespace: istio-system
+spec:
+  ...
+  gateways:
+  - api-gateway
+  http:
+  - match:
+    - method:
+        exact: CONNECT
+      headers:
+        X-Gardener-API-Route:
+          regex: ".*"
+    ...
+```
+
+This configuration sets up the Istio IngressGateway to accept HTTPS traffic and route HTTP CONNECT requests with the proposed custom header to the `kube-apiserver`.
+
+### EnvoyFilter for Custom Header Processing
+
+To process the proposed custom header and make routing decisions based on it, it will be required to additionally add and/or modify an EnvoyFilter. Again, existing configuration may be reused adjusted to this new use case to keep convention parity within the code base.
+
+```yaml
+{{- if eq .Values.vpn.enabled true -}}
+apiVersion: networking.istio.io/v1beta1
+kind: Gateway
+metadata:
+  name: reversed-vpn-auth-server
+  namespace: {{ .Release.Namespace }}
+spec:
+  selector:
+{{ .Values.labels | toYaml | indent 4 }}
+  servers:
+  - hosts:
+    - reversed-vpn-auth-server.garden.svc.cluster.local
+    port:
+      name: tls-tunnel
+      number: 8134
+      protocol: HTTP
+{{- end }}
+```
+### Implementation Steps
+
+1. Apply the Istio Gateway and VirtualService configurations.
+2. Deploy the EnvoyFilter for custom header processing.
+3. Update any relevant Istio authorization policies to allow the new traffic flow.
+4. Update the `apiserver-proxy` code or configuration to use HTTP CONNECT and include the custom header.
+5. Deploy the updated `apiserver-proxy` configuration.
+
+## Phased Rollout
+
+The implementation will be controlled by two distinct feature gates that handle different aspects of the solution. Each feature gate can be disabled per shoot to allow testing of the old implementation via E2E tests.
+
+### Feature Gate 1: `APIServerSecureRouting`
+Controls whether shoots use the new secure routing implementation with HTTP CONNECT and custom headers.
+
+**Rollout Plan:**
+- Initially introduced as disabled by default
+- Once stability is proven in production environments, enabled by default for new shoots
+- Existing shoots will retain their previous setting until explicitly migrated
+- When fully proven, the feature gate will be removed and the functionality will become permanent
+
+### Feature Gate 2: `APIServerLegacyPortDisable`
+Controls whether the legacy port is available. This feature gate can only be enabled after `APIServerSecureRouting` has been fully rolled out.
+
+**Rollout Plan:**
+- Introduced as disabled by default
+- Once all shoots have migrated to secure routing, enabled by default for new shoots
+- Existing shoots will be notified to migrate during reconciliation
+- When migration is complete, the feature gate will be removed and the legacy port will be permanently disabled
```

Comment:
I do not get why we need this feature gate at all. Also, I see no reason to notify anybody for a migration. This should be done automatically from my point of view.